### ISSUE TITLE
Getting gradient of LRP with respect to the input

### ISSUE THREAD

**Issue by MikiFER** [pytorch/captum#1025]
Hi,

Is there a way to obtain a gradient of the input relevance heatmap obtained by LRP with respect to the input?

**Comment by NarineK** [comment]
@MikiFER, current implementation of LRP returns the relevance scores w.r.t. input that you can visualize as a heatmap ? Are you looking for something else ? 
https://github.com/pytorch/captum/blob/master/captum/attr/_core/lrp.py#L231

**Comment by MikiFER** [comment]
Hi @NarineK first of all great job on implementing the library and thank you for responding.

Yes, what I need for my research is to optimize model's weights. That is I need computational graph of the heatmap w.r.t. to the input so I can back-propagate through it and modify model's parameters. I tested captum and saw that it returns a vector which has a gradient_fn parameter and then I visualized the graph with torchviz.

The only difference between the graph obtained with the normal forward pass compared to the one obtained by running the tool on the returned heatmap were 4 additional nodes (SliceBackward0, SelectBackward0, ReshapeAliasBackward0 and MulBackward0) even though in my opinion there should be a lot more because in the backward pass to obtain LRP heatmap there are multiple multiplications of the model's weights with the activation in order to obtain the heatmap (graph should be twice as long).

**Comment by craymichael** [comment]
Apologies for the delay.

Captum's LRP implementation is designed to return relevance scores, not to provide a fully differentiable computation graph through the relevance propagation itself.

For methods such as Saliency, it is possible to customize the gradient function to call `torch.autograd.grad(..., create_graph=True)` and then differentiate the attribution again. LRP is harder because it uses forward / backward hooks and modifies gradient propagation rules during the backward pass. That path is not currently a stable higher-order differentiable objective for optimizing model parameters.

If your goal is to train with an attribution regularizer, I would recommend starting with a gradient method that can be made higher-order differentiable, such as Saliency / InputXGradient with `create_graph=True`, rather than LRP.

This should answer the question, so I'll close this for now. Please feel free to reply with any follow-up questions.
